dev-docs/RFCs/proposals/transport-format-rfc.md

```diff
@@ -0,0 +1,110 @@
+# RFC: Transport Format RFC
+
+* Authors: Xiaoji Chen
+* Date: Jan, 2020
+* Status: **Draft**
+
+## Summary
+
+This RFC proposes a generic transport format for generating/rendering data cross platforms.
+
+## Background
+
+In v7.3, deck.gl added the `@deck.gl/json` module. It supports generating serializable deck.gl props in non-JavaScript environments and send it to render on the client side. This format is currently used by `pydeck`, deck.gl's Python binding.
+
+We have plans to add bindings for other languages, and potentially non-JS renderers. This calls for a more comprehensive format that can be easily generated, sent, and handled by multiple languages and platforms.
+
+The JSON format has the following shortcomings:
+
+- Efficiency: Serializing/deserializing large data objects with JSON is expensive in terms of both time and space. Network bandwidth, CPU time and memory are all valuable assets when it comes to web applications.
+- Updates: It is not possible to update a layer without re-sending the entire snapshot. This is important for animation and/or live data visualization.
+
+A hybrid transport format of both JSON and binary will help address these issues.
+
+## Proposal
+
+Expand the `JSONConverter.convert` API to accept a second argument:
+
+```js
+converter.convert(json, buffers);
```

Comment:
There is not much detail on how buffers are transported, which is somewhat reasonable given that we may have widely varying implementations on different platforms. 

But we probably need a buffer or data source manager (perhaps on the @deck.gl/json level) and it probably needs to have a defined API that the implementations can work against.

They requirement is of course that buffers can be sent separately from JSON and that JSON can be updated multiple times, referencing the same buffers.

Reply:
```They requirement is of course that buffers can be sent separately from JSON and that JSON can be updated multiple times, referencing the same buffers.```

This part seems to be an important requirement not really covered here.  The idea is that there is a set of state (JSON and Buffers) and you can update both or one, and the "correct" behavior should happen.

Does this also mean supporting partial updates?  For example, I want to only update 1 buffer out of the 10 that where sent.

Reply:
Yes look at the message sequence diagram in the linked slides. Ideally there should be the ability to send independent messages with data "blobs", and independent messages with JSON styles. 

I am envisioning e.g. a pydeck user working interactively with pyarrow, creating binary blobs, sending them to deck, then sending new JSON layers  referencing those new blobs to update the animation via e.g. transitions.

And yes, for that kind of use case it would be cool if "named" data blobs could be replaced (or as you say "partially updated", i.e. not just added.